javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -155,57 +211,35 @@ T get(int index) {
     /**
      * fetch the leaf, corresponding to the given index.
      * Node: the offset and length should be taken into consideration as there may be leading and trailing garbage.
+     * Also, the returned array is mutable, but should not be mutated!
      */
     @SuppressWarnings("unchecked")
     T[] getLeaf(int index) {
-        Object[] leaf = array;
-        for (int shift = depthShift; shift > 0; shift -= BRANCHING_BASE) {
-            leaf = getAt(leaf, digit(offset + index, shift));
-        }
-        return (T[]) leaf;
-    }
-
-    Iterator<T> iterator() {
-        return new Iterator<T>() {
-            final int globalLength = BitMappedTrie.this.length();
-            int globalIndex = 0;
-
-            int index = lastDigit(offset());
-            T[] leaf = getLeaf(globalIndex);
-            int length = leaf.length;
-
-            @Override
-            public boolean hasNext() { return globalIndex < globalLength; }
-
-            @Override
-            public T next() {
-                if (index == length) { setCurrentArray(); }
-
-                final T next = leaf[index];
-
-                index++;
-                globalIndex++;
-
-                return next;
-            }
-
-            void setCurrentArray() {
-                index = 0;
-                leaf = getLeaf(globalIndex);
-                length = leaf.length;
-                assert leaf != null;
+        if (depthShift == 0) {
+            return (T[]) array;
+        } else if (depthShift == BRANCHING_BASE) {
+            return getAt(array, firstDigit(offset + index, depthShift));
+        } else {
+            index += offset;
+            Object[] leaf = getAt(array, firstDigit(index, depthShift));
+            for (int shift = depthShift - BRANCHING_BASE; shift > 0; shift -= BRANCHING_BASE) {
+                leaf = getAt(leaf, digit(index, shift));
             }
-        };
+            return (T[]) leaf;
+        }
     }
 
-    int offset() { return offset; }
-    int length() { return length; }
+    int offset()    { return offset; }
+    int length()    { return length; }
+    int leafCount() { return (int) Math.ceil(length() / BRANCHING_FACTOR); }
 }
 
 @FunctionalInterface
 interface NodeModifier {
     Object[] apply(Object arrayObject, int index);
 
-    static NodeModifier copyLeaf = (o, i) -> copy(o, i + 1);
-    static NodeModifier updateLeafWith(Object element) { return (o, i) -> copyUpdate(o, i, element); }
+    static <T> NodeModifier updateLeafWith(T element)  { return (o, i) -> copyUpdate(o, i, element); }
+    static NodeModifier replaceLeafWith(Object[] leaf) { return (o, i) -> leaf; }
+    NodeModifier copyNode = (o, i) -> copy(o, i + 1);
+    NodeModifier Identity = (o, i) -> (Object[]) o;
```

Comment:
These interface fields are implicitly `static final`, i.e. constants. By naming convention the names have to be upper-case:

```java
NodeModifier COPY_NODE = (o, i) -> copy(o, i + 1);
NodeModifier IDENTITY = (o, i) -> (Object[]) o;
```